File: src/ast/SpecialSubstitution.java
package ast;

import java.io.StringWriter;

public class SpecialSubstitution extends BaseNode {
    public enum SpecialType {
        Allocator,
        BasicString,
        String,
        IStream,
        OStream,
        IOStream
    }

    private SpecialType _specialSubstitutionKey;

    public SpecialSubstitution(SpecialType specialSubstitutionKey) {
        super(NodeType.SpecialSubstitution);
        _specialSubstitutionKey = specialSubstitutionKey;
    }

    public void SetExtended() {
        Type = NodeType.ExpandedSpecialSubstitution;
    }

    @Override
    public String GetName() {
        switch (_specialSubstitutionKey) {
            case Allocator:
                return "allocator";
            case BasicString:
                return "basic_string";
            case String:
                if (Type == NodeType.ExpandedSpecialSubstitution) {
                    return "basic_string";
                }

                return "string";
            case IStream:
                return "istream";
            case OStream:
                return "ostream";
            case IOStream:
                return "iostream";
        }

        return null;
    }

    //No override
    private String GetExtendedName() {
        switch (_specialSubstitutionKey) {
            case Allocator:
                return "std::allocator";
            case BasicString:
                return "std::basic_string";
            case String:
                return "std::basic_string<char, std::char_traits<char>, std::allocator<char> >";
            case IStream:
                return "std::basic_istream<char, std::char_traits<char> >";
            case OStream:
                return "std::basic_ostream<char, std::char_traits<char> >";
            case IOStream:
                return "std::basic_iostream<char, std::char_traits<char> >";
        }

        return null;
    }

    @Override
    public void PrintLeft(StringWriter writer) {
        if (Type == NodeType.ExpandedSpecialSubstitution) {
            writer.write(GetExtendedName());
        } else {
            writer.write("std::");
            writer.write(GetName());
        }
    }
}

File: src/de/fearlesstobi/demangler/ast/SpecialSubstitution.kt
package de.fearlesstobi.demangler.ast

import java.io.StringWriter

class SpecialSubstitution(private val specialSubstitutionKey: SpecialType) : BaseNode(NodeType.SpecialSubstitution) {
    enum class SpecialType {
        Allocator, BasicString, String, IStream, OStream, IOStream
    }

    fun SetExtended() {
        type = NodeType.ExpandedSpecialSubstitution
    }

    override fun getName(): String? {
        return when (specialSubstitutionKey) {
            SpecialType.Allocator -> "allocator"
            SpecialType.BasicString -> "basic_string"
            SpecialType.String -> {
                if (type == NodeType.ExpandedSpecialSubstitution) {
                    "basic_string"
                } else "string"
            }
            SpecialType.IStream -> "istream"
            SpecialType.OStream -> "ostream"
            SpecialType.IOStream -> "iostream"
        }
        return null
    }

    //No override
    private fun GetExtendedName(): String {
        return when (specialSubstitutionKey) {
            SpecialType.Allocator -> "std::allocator"
            SpecialType.BasicString -> "std::basic_string"
            SpecialType.String -> "std::basic_string<char, std::char_traits<char>, std::allocator<char> >"
            SpecialType.IStream -> "std::basic_istream<char, std::char_traits<char> >"
            SpecialType.OStream -> "std::basic_ostream<char, std::char_traits<char> >"
            SpecialType.IOStream -> "std::basic_iostream<char, std::char_traits<char> >"
        }
        return ""
    }

    public override fun printLeft(writer: StringWriter) {
        if (type == NodeType.ExpandedSpecialSubstitution) {
            writer.write(GetExtendedName())
        } else {
            writer.write("std::")
            writer.write(getName())
        }
    }

}